operator/pkg/controllers/configconnectorcontext/experiments.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configconnectorcontext
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative/pkg/manifest"
+
+	corev1beta1 "github.com/GoogleCloudPlatform/k8s-config-connector/operator/pkg/apis/core/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/operator/pkg/experiments"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func (r *Reconciler) transformForExperiments(kube client.Client) declarative.ObjectTransform {
+	return func(ctx context.Context, o declarative.DeclarativeObject, m *manifest.Objects) error {
+		ccc, ok := o.(*corev1beta1.ConfigConnectorContext)
+		if !ok {
+			return fmt.Errorf("expected the resource to be a ConfigConnector, but it was of type %T", o)
+		}
+
+		cc := &corev1beta1.ConfigConnector{}
+		ccName := client.ObjectKey{
+			Namespace: "", Name: corev1beta1.ConfigConnectorAllowedName,
+		}
+		if err := kube.Get(ctx, ccName, cc); err != nil {
+			return fmt.Errorf("getting cluster ConfigConnector object %v: %w", ccName, err)
+		}
+
+		if err := r.applyExperiments(ctx, cc, ccc, m); err != nil {
+			return fmt.Errorf("error applying experiment transforms: %w", err)
+		}
+		return nil
+	}
+}
+
+func (r *Reconciler) applyExperiments(ctx context.Context, cc *corev1beta1.ConfigConnector, ccc *corev1beta1.ConfigConnectorContext, m *manifest.Objects) error {
+	log := log.FromContext(ctx)
+
+	for _, experiment := range cc.Spec.Experiments {
+		key := strings.ToLower(experiment)
+		switch key {
+		case experiments.LegacyIAMReconciler:
+			if err := r.applyExperimentLegacyIAMReconciler(ctx, m); err != nil {
+				return err
+			}
+
+		default:
+			log.Info("ignoring unknown experiment", "key", key)
+			// TODO: add to status?
+		}
+	}
+
+	return nil
+}
+
+func (r *Reconciler) applyExperimentLegacyIAMReconciler(ctx context.Context, m *manifest.Objects) error {
+	log := log.FromContext(ctx)
+	for _, obj := range m.Items {
+		if obj.Kind != "StatefulSet" || !strings.HasPrefix(obj.GetName(), "cnrm-controller-manager") {
```

Comment:
Can we document the behavior of the experiment you are adding? I am not sure I fully understand the logic here. Do we want to be able to scope the experiment to namespace scope? Or is the experiment always cluster scoped?

Reply:
Oh actually we might have misread the code earlier. I think the idea is that if KCC is running in cluster mode, use CC's spec.experiments field; If KCC is running in namespaced mode, use CCC's spec.experiments field?